src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+ * to be applied to a single module.
+ */
+export class ModulePluginWrapper implements PluginWrapper {
+  // Sentinel value to indicate that a plugin has not been loaded into memory
+  // yet.
+  private static readonly NOT_LOADED: Plugin = [];
+  private unpatchFns: Array<() => void> = [];
+  // A logger.
+  private logger: Logger;
+  // Configuration for a TraceAgent instance.
+  private traceConfig: TraceAgentConfig;
+  // Display-friendly name of the module being patched by this plugin.
+  private name: string;
+  // The path to the plugin.
+  private path: string;
+  // The exported value of the plugin, or NOT_LOADED if it hasn't been
+  // loaded yet.
+  private pluginExportedValue: Plugin = ModulePluginWrapper.NOT_LOADED;
+  private traceApiInstances: TraceAgent[] = [];
 
-function isInternalIntercept<T>(obj: InternalInstrumentation<T>):
-    obj is InternalIntercept<T> {
-  return !!(obj as InternalIntercept<T>).intercept;
-}
+  /**
+   * Constructs a new PluginWrapper instance.
+   * @param logger The logger to use.
+   * @param options Initialization fields for this object.
+   * @param traceConfig Configuration for a TraceAgent instance.
+   */
+  constructor(
+      logger: Logger, options: ModulePluginWrapperOptions,
+      traceConfig: TraceAgentConfig) {
+    this.logger = logger;
+    this.name = options.name;
+    this.path = options.path;
+    this.traceConfig = traceConfig;
+  }
 
-let plugins: PluginStore = Object.create(null);
-// tslint:disable-next-line:no-any
-let intercepts: {[moduleName: string]: {interceptedValue: any}} =
-    Object.create(null);
-let activated = false;
-// TODO(kjin): Make plugin loader a singleton, so we can avoid shadowing.
-// tslint:disable-next-line:variable-name
-let logger_: Logger;
-
-function checkLoadedModules(): void {
-  for (const moduleName of Object.keys(plugins)) {
-    // \\ is benign on unix and escapes \\ on windows
-    const regex =
-        new RegExp('node_modules\\' + path.sep + moduleName + '\\' + path.sep);
-    for (const file of Object.keys(require.cache)) {
-      if (file.match(regex)) {
-        logger_.error(
-            moduleName + ' tracing might not work as ' + file +
-            ' was loaded before the trace agent was initialized.');
-        break;
+  isSupported(version: string): boolean {
+    // The plugin is lazily loaded here.
+    const plugin = this.getPluginExportedValue();
+    // Count the number of Patch/Intercept objects with compatible version
+    // ranges
+    let numFiles = 0;
+    plugin.forEach(instrumentation => {
+      const postLoadVersions = instrumentation.versions;
+      if (!postLoadVersions || semver.satisfies(version, postLoadVersions)) {
+        numFiles++;
       }
+    });
+    // We consider a module to be unsupported if there are no Patch/Intercept
+    // objects with compatible version ranges at all.
+    return numFiles > 0;
+  }
+
+  unapplyAll() {
+    this.unpatchFns.reverse().forEach(fn => fn());
```

Comment:
It's probably worthwhile to add a comment that unpatch should be done in the reverse order.